### PR TITLE
Refactor/`embed.BaseCnn` address issue #513

### LINKED ISSUE #513: [Refactor] Introduce `BaseCNN` in `cnn.py` and Refactor Existing CNN Classes under the `embed` API
Introduce a reusable `BaseCNN` class to encapsulate shared functionality across the CNN variants under the `embed` API. Refactor the existing CNN implementations to inherit from this base class, improving code maintainability and reducing redundancy.

- Modules that will be modified: `embed.cnn`, `embed.image_cnn`, and `embed.signal_cnn`
- Impact on code structure: these modules may be consolidated into a single implementation
- Will this be a breaking change? **Yes**

### PR DESCRIPTION
**Pull request by cmanjah**
Fixes #513.

### Description

This PR creates a `BaseCNN` base class that consolidates common CNN utilities and refactors 7 CNN classes (CNNEncoder, ProteinCNN, SignalVAEEncoder, LeNet, ImageVAEEncoder, SmallCNNFeature, SimpleCNNBuilder) to inherit from it. This eliminates ~500 lines of duplicated code (~40% reduction) while improving maintainability and ensuring FAIR compliance.

**Key changes:**
- Created BaseCNN with 10 reusable utility methods (_apply_activation, _create_sequential_conv_blocks, _create_doubling_conv_blocks, _create_progressive_conv_blocks, _initialize_weights, _flatten_features, _apply_pooling, _create_embedding_layer, _get_conv_output_size, and __init__)
- Refactored 7 CNN classes across 3 files (cnn.py, signal_cnn.py, image_cnn.py) to use BaseCNN utilities
- Added FAIR compliance methods (output_size(), __repr__()) to all refactored classes
- Improved forward() methods to use iteration pattern for better maintainability
- Added comprehensive Google-style docstrings to all methods (100% documentation coverage)
- Maintained 100% backward compatibility (preserved conv1/conv2/conv3 attributes)

All 46 tests passing. Pre-commit hooks (flake8, black, isort, mypy) passing.

### Status
**Ready**

### Types of changes
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] In-line docstrings updated.
- [ ] [Source for documentation at `docs`](https://github.com/pykale/pykale/tree/main/docs/source) manually updated for new API.

### CODE REVIEW

**Comment by haipinglu** [comment]
@codex review

**Comment by haipinglu** [comment]
@codex review the full PR diff (all files changed), and report the top 5 Must-fix/Important issues across the whole PR

File: kale/embed/cnn.py [hunk 1]
```diff
@@ -195,7 +313,7 @@ def __init__(
         height = cnn_output_shape[2]
         width = cnn_output_shape[3]
 
-        encoder_layer = nn.TransformerEncoderLayer(num_channels, num_heads, dim_feedforward, dropout)
+        encoder_layer = nn.TransformerEncoderLayer(num_channels, num_heads, dim_feedforward, dropout, batch_first=True)
```
[the remark this answers is not in the document]
Comment (cmanjah) [reply kale/embed/cnn.py]: Removed `batch_first=True` from `CNNTransformer` to maintain consistency with `spatial_to_seq` and `PositionalEncoding`

File: kale/embed/cnn.py [hunk 2]
```diff
@@ -58,29 +139,66 @@ class ProteinCNN(nn.Module):
...
+        padding_idx = 0 if padding else None
+        self.embedding = self._create_embedding_layer(
+            num_embeddings=26, embedding_dim=embedding_dim, padding_idx=padding_idx
+        )
+
+        conv_layers, batch_norms = self._create_sequential_conv_blocks(
+            in_channels=embedding_dim,
+            out_channels_list=num_filters,
+            kernel_sizes=kernel_size,
+            conv_type="1d",
```
[the remark this answers is not in the document]
Comment (cmanjah) [reply kale/embed/cnn.py:155]: Added explicit `paddings=0` to `ProteinCNN` to preserve original no-padding behavior

File: kale/embed/image_cnn.py [hunk 3]
```diff
@@ -413,22 +658,45 @@ class ImageVAEEncoder(nn.Module):
...
-        self.conv3 = nn.Conv2d(32, 64, kernel_size=3, stride=2, padding=1)
+
+        self.conv_layers, _ = self._create_sequential_conv_blocks(
+            in_channels=input_channels,
+            out_channels_list=[16, 32, 64],
+            kernel_sizes=[3, 3, 3],
+            conv_type="2d",
+            strides=[2, 2, 2],
+            paddings=[1, 1, 1],
+            use_batch_norm=False,
```
[the remark this answers is not in the document]
Comment (cmanjah) [reply kale/embed/image_cnn.py:726]: Added `bias=True` to `ImageVAEEncoder` and `SmallCNNFeature` for checkpoint compatibility

File: kale/embed/image_cnn.py [hunk 4]
```diff
@@ -387,23 +606,49 @@ def forward(self, x):
...
 
-class ImageVAEEncoder(nn.Module):
+        Returns:
+            int: Number of output channels in the final convolutional layer.
+                This is (2^(num_layers-1)) * base_channels.
+        """
+        num_layers = len(self.conv_layers)
+        if num_layers > 0:
+            # Last layer has channels: (2^(num_layers-1)) * base_channels
+            return self.conv_layers[-1].out_channels
```
[the remark this answers is not in the document]
Comment (cmanjah) [reply kale/embed/image_cnn.py:651]: Fixed `LeNet.output_size()` to correctly return `linear.out_features` when linear layer is configured

File: kale/embed/image_cnn.py [hunk 5]
```diff
@@ -47,49 +52,99 @@ def forward(self, x):
...
+    def __init__(self, num_channels: int = 3, kernel_size: int = 5):
         super(SmallCNNFeature, self).__init__()
-        self.conv1 = nn.Conv2d(num_channels, 64, kernel_size=kernel_size)
-        self.bn1 = nn.BatchNorm2d(64)
+
+        # Use padding=0 (no padding) to match original behavior that reduces spatial dimensions
+        self.conv_layers, self.batch_norms = self._create_sequential_conv_blocks(
+            in_channels=num_channels,
+            out_channels_list=[64, 64, 128],
+            kernel_sizes=[kernel_size, kernel_size, kernel_size],
```
[the remark this answers is not in the document]
Comment (cmanjah) [reply kale/embed/image_cnn.py:83]: Restored `conv1/conv2/conv3` and `bn1/bn2/bn3` aliases for backward API compatibility

File: kale/embed/cnn.py [hunk 6]
```diff
@@ -58,29 +139,67 @@ class ProteinCNN(nn.Module):
...
+            paddings=0,
+            use_batch_norm=True,
+            bias=True,
+        )
+
+        # Maintain backward compatibility: expose individual layer attributes
+        self.conv1 = conv_layers[0]
```
Comment (shuo-zhou) [inline kale/embed/cnn.py]: Would it be better to use two lists `self.conv` and `self.bn`?

File: kale/embed/cnn.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shuo-zhou) [reply kale/embed/cnn.py:155]: The `padding` argument in `ProteinCNN` may be confused with `paddings`.

File: kale/embed/cnn.py [hunk 7]
```diff
@@ -58,29 +139,67 @@ class ProteinCNN(nn.Module):
 
     def __init__(self, embedding_dim, num_filters, kernel_size, padding=True):
```
Comment (shuo-zhou) [inline kale/embed/cnn.py]: `padding: bool = True`

File: kale/embed/image_cnn.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (shuo-zhou) [reply kale/embed/image_cnn.py:83]: We may keep it simple in future updates.

File: kale/embed/image_cnn.py [hunk 8]
```diff
@@ -99,40 +163,86 @@ class SimpleCNNBuilder(nn.Module):
...
             Defaults to (0,3). (0,3) means placing 2 pooling layers after the first and fourth convolutional layer.
-        num_channels (int): the number of input channels. Defaults to 3.
-    """
 
-    activations = {"relu": nn.ReLU(), "elu": nn.ELU(), "leaky_relu": nn.LeakyReLU()}
+    Example:
+        >>> # Build a CNN with custom layer specifications
+        >>> model = SimpleCNNBuilder(
+        ...     conv_layers_spec=[[16, 3], [32, 3], [64, 3]],
+        ...     activation_fun='relu',
```
Comment (shuo-zhou) [inline kale/embed/image_cnn.py]: `activation_func` is better for me. Please check all.

File: kale/embed/image_cnn.py [hunk 9]
```diff
@@ -99,40 +163,86 @@ class SimpleCNNBuilder(nn.Module):
...
     def __init__(
         self, conv_layers_spec, activation_fun="relu", use_batchnorm=True, pool_locations=(0, 3), num_channels=3
     ):
         super(SimpleCNNBuilder, self).__init__()
-        self.layers = nn.ModuleList()
+        self.conv_layers_spec = conv_layers_spec
+        self.activation_fun = activation_fun
+        self.use_batchnorm = use_batchnorm
+        self.pool_locations = pool_locations
+        self.num_input_channels = num_channels
```
Comment (shuo-zhou) [inline kale/embed/image_cnn.py]: It would be better to keep consistency; either `num_input_channels` or `num_channels` is fine.

File: kale/embed/image_cnn.py [hunk 10]
```diff
@@ -413,22 +670,51 @@ class ImageVAEEncoder(nn.Module):
...
+        """
         super().__init__()
-        # Convolutional layers for 224x224 input
-        self.conv1 = nn.Conv2d(input_channels, 16, kernel_size=3, stride=2, padding=1)
-        self.conv2 = nn.Conv2d(16, 32, kernel_size=3, stride=2, padding=1)
-        self.conv3 = nn.Conv2d(32, 64, kernel_size=3, stride=2, padding=1)
+
+        self.conv_layers, _ = self._create_sequential_conv_blocks(
+            in_channels=input_channels,
+            out_channels_list=[16, 32, 64],
```
Comment (shuo-zhou) [inline kale/embed/image_cnn.py]: `out_channels_size_list` may be better. Check all

File: kale/embed/image_cnn.py [hunk 11]
```diff
@@ -413,22 +670,51 @@ class ImageVAEEncoder(nn.Module):
...
+        (e.g., adjust the linear layer input).
     """
 
-    def __init__(self, input_channels=1, latent_dim=256):
+    def __init__(self, input_channels: int = 1, latent_dim: int = 256):
+        """
+        Initialize the ImageVAEEncoder model.
+
+        Args:
+            input_channels (int): Number of input channels (default=1).
```
Comment (shuo-zhou) [inline kale/embed/image_cnn.py]: It would be better to keep the argument name consistent. Use `num_channels` or `num_input_channels`.

**Review by shuo-zhou — CHANGES_REQUESTED** [review]
Looks good overall. I left a few minor comments for you to address.

File: kale/embed/image_cnn.py (hunk 9 above)
Comment (cmanjah) [reply kale/embed/image_cnn.py]: Done! Standardized on `num_channels` across:
- image_cnn.py: SmallCNNFeature, SimpleCNNBuilder, LeNet, ImageVAEEncoder
- cnn.py: CNNEncoder
- signal_cnn.py: Signal1DCNN

All classes now consistently use `num_channels` for input channel count, with deprecated aliases for backward compatibility.

File: kale/embed/image_cnn.py (hunk 10 above)
Comment (cmanjah) [reply kale/embed/image_cnn.py]: Renamed `out_channels_list` → `out_channels_size_list` everywhere (**19 occurrences**)

File: kale/embed/image_cnn.py (hunk 8 above)
Comment (cmanjah) [reply kale/embed/image_cnn.py]: `SimpleCNNBuilder` now uses `activation_func` in the _signature/docs/examples_; `activation_fun` is supported as a deprecated alias for backward compatibility.

File: kale/embed/image_cnn.py (hunk 11 above)
Comment (cmanjah) [reply kale/embed/image_cnn.py]: Standardized to `num_channels` across `SmallCNNFeature`, `SimpleCNNBuilder`, `LeNet`, and `ImageVAEEncoder`. Added deprecated `input_channels` aliases for `LeNet `and `ImageVAEEncoder` to preserve backward compatibility.

File: kale/embed/cnn.py (hunk 6 above)
Comment (cmanjah) [reply kale/embed/cnn.py]: Created `self.conv_layers` and `self.batch_norms` as `nn.ModuleList` attributes in both `CNNEncoder` and `ProteinCNN`, while keeping backward-compatible individual aliases.

File: kale/embed/cnn.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cmanjah) [reply kale/embed/cnn.py:155]: Renamed parameter to `use_padding` with enhanced documentation and backward-compatible alias.

File: kale/embed/image_cnn.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cmanjah) [reply kale/embed/image_cnn.py:83]: Noted—we can revisit this in a future cleanup pass

File: kale/embed/cnn.py (hunk 7 above)
Comment (cmanjah) [reply kale/embed/cnn.py]: `padding` here is a boolean flag that enables the default PyTorch padding behavior. I’ve kept the name to stay consistent with the underlying API, but I’m happy to revisit naming in a future cleanup.

**Comment by shuo-zhou** [comment]
> **@shuo-zhou**, The job failed due to a ConnectTimeout error when trying to download data from https://dataverse.harvard.edu/api/access/datafile/4291555 in [/examples/bindingdb_deepdta/tutorial.ipynb](https://github.com/pykale/pykale/blob/main/examples/bindingdb_deepdta/tutorial.ipynb).
> 
> **Root Cause:** The notebook [/examples/bindingdb_deepdta/tutorial.ipynb](https://github.com/pykale/pykale/blob/main/examples/bindingdb_deepdta/tutorial.ipynb) depends on downloading a data file from dataverse.harvard.edu, but the connection timed out (MaxRetryError). The rest of the tests passed—this error is isolated to the external data fetch.
> 
> **Solution:** The diagnosis (connection timeout to external data source) Suggested fix (see code below) Do you prefer to cache this data artifact or mark the test as "allowed to fail"?
> 
> ```
> import requests
> from time import sleep
> 
> def download_with_retry(url, output_path, retries=3, timeout=120):
>     """Download file with retry logic for flaky networks."""
>     for attempt in range(retries):
>         try:
>             response = requests.get(url, timeout=timeout)
>             response.raise_for_status()
>             with open(output_path, "wb") as f:
>                 f.write(response.content)
>             print(f"Downloaded to {output_path}")
>             return True
>         except requests.exceptions.RequestException as e:
>             if attempt < retries - 1:
>                 wait_time = 2 ** attempt  # exponential backoff
>                 print(f"Download failed (attempt {attempt+1}/{retries}): {e}")
>                 print(f"Retrying in {wait_time}s...")
>                 sleep(wait_time)
>             else:
>                 print(f"Failed to download after {retries} attempts: {e}")
>                 print("Please download manually from: {url}")
>                 return False
> 
> # Use it
> url = "https://dataverse.harvard.edu/api/access/datafile/4291555"
> success = download_with_retry(url, "data_file.pkl")
> if not success:
>     raise RuntimeError("Data download failed. Cannot proceed with tutorial.")
> ```

@Cmanjah Please create an issue and a new PR for this.

**Review by shuo-zhou — APPROVED** [review]
LGTM. I created issue #518 to revisit the naming discussions.

File: tests/embed/test_cnn.py [hunk 12]
```diff
@@ -50,9 +77,354 @@ def test_protein_cnn_forward():
...
+    """Test CNNEncoder output_size method."""
+    model = CNNEncoder(num_embeddings=100, embedding_dim=128, sequence_length=1000, num_kernels=32, kernel_length=4)
+    output_size = model.output_size()
+    assert isinstance(output_size, int) or isinstance(output_size, tuple)
+    # CNNEncoder returns (96,) as tuple
+    if isinstance(output_size, tuple):
+        assert len(output_size) == 1
+        assert output_size[0] == 96  # num_kernels * 3
+    else:
+        assert output_size == 96
```
[the remark this answers is not in the document]
Comment (cmanjah) [reply tests/embed/test_cnn.py:108]: Not necessary

File: kale/embed/cnn.py [hunk 13]
```diff
@@ -195,7 +321,7 @@ def __init__(
         height = cnn_output_shape[2]
         width = cnn_output_shape[3]
 
-        encoder_layer = nn.TransformerEncoderLayer(num_channels, num_heads, dim_feedforward, dropout)
+        encoder_layer = nn.TransformerEncoderLayer(num_channels, num_heads, dim_feedforward, dropout, batch_first=True)
```
[the remark this answers is not in the document]
Comment (cmanjah) [reply kale/embed/cnn.py]: Removed `batch_first=True`

**Review by shuo-zhou — APPROVED** [review]
LGTM
